src/rust_connection/mod.rs

```diff
@@ -262,9 +248,86 @@ impl<R: ReadFD, W: WriteFD> RustConnection<R, W> {
             .send_request(ReplyFDKind::ReplyWithoutFDs)
             .expect("Sending a HasResponse request should not be blocked by syncs");
         inner.discard_reply(seqno, DiscardMode::DiscardReplyAndError);
-        write_all_vectored(write, &[IoSlice::new(&request)], Vec::new())?;
+        let inner = self.write_all_vectored(inner, write, &[IoSlice::new(&request)], Vec::new())?;
 
-        Ok(())
+        Ok(inner)
+    }
+
+    /// Write a set of buffers on a `writer`. May also read packets
+    /// from the server.
+    fn write_all_vectored<'a>(
+        &'a self,
+        mut inner: MutexGuardInner<'a>,
+        write: &mut W,
+        mut bufs: &[IoSlice<'_>],
+        mut fds: Vec<RawFdContainer>,
+    ) -> std::io::Result<MutexGuardInner<'a>> {
+        let mut partial_buf: &[u8] = &[];
+        while !partial_buf.is_empty() || !bufs.is_empty() || !fds.is_empty() {
+            let (can_read, can_write) = write.poll(true, true)?;
+            if can_write {
+                let mut count = if !partial_buf.is_empty() {
+                    write.write(partial_buf, &mut fds)?
+                } else {
+                    write.write_vectored(bufs, &mut fds)?
+                };
+                if count == 0 {
+                    return Err(std::io::Error::new(
+                        std::io::ErrorKind::WriteZero,
+                        "failed to write anything",
+                    ));
+                }
+                if count >= partial_buf.len() {
+                    count -= partial_buf.len();
+                    partial_buf = &[];
+                } else {
+                    partial_buf = &partial_buf[count..];
+                    count = 0;
+                }
+                while count > 0 {
+                    if count >= bufs[0].len() {
+                        count -= bufs[0].len();
+                    } else {
+                        partial_buf = &bufs[0][count..];
```

Comment:
Unrelated to this PR, but: should we change the request sending APIs to get a `&mut [IoSlice]` instead of just `&[IoSlice]`? That would allow to get rid of some of this ugly partial buffer write handling.
If so, we should open an issue about this idea. At least that's my usual approach for this kind of ideas. :-)
(Perhaps there even is already such an issue and I just forgot about it?)

Reply:
Yes, let's open an issue and think about it later.